sql/sem/builtins: add crdb_internal.datums_to_bytes, use in hash sharding

ajwerner opened the pull request:
This new builtin function encodes datums using the key encoding. It is useful
in particular because it is a function with immutable volatility from a large
number of data types to bytes.

The function is then adopted as the default expression for hash sharded indexes if the version is new enough.

Fixes #67170.

Release note (sql change): Added a builtin function,
`crdb_internal.datums_to_bytes`, which can encode any data type which can be
used in an forward index key into bytes in an immutable way. This function
is now used in the expression for hash sharded indexes.

ajwerner committed:
sql/sem/builtins: add crdb_internal.datums_to_bytes for hash sharding

This new builtin function encodes datums using the key encoding. It is useful
in particular because it is a function with immutable volatility from a large
number of data types to bytes.

Release note (sql change): Added a builtin function,
crdb_internal.datums_to_bytes, which can encode any data type which can be
used in an forward index key into bytes in an immutable way. This function
is now used in the expression for hash sharded indexes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67865)
<!-- Reviewable:end -->

ajwerner commented:
@RaduBerinde this PR has the changes. Ideally that big select in `execbuilder/testdata/hash_sharded_index` would be a point lookup.

RaduBerinde reviewed (commented):
Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-:-MfBhbDE-ZB874s-yvsH:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/create_table.go, line 2716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhnzQ-dhs_EYUS0au:-MfBhnzQ-dhs_EYUS0av:br6z81o) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2716)):*
> ```Go
> }
> 
> func makeHashShardComputeExpr(colNames []string, buckets int) *string {
> ```

[nit] probably on your TODO list but this should have a comment with sample expressions
___
*[pkg/sql/create_table.go, line 2752 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi2dZ-8FuBQFyrkhx:-MfBi2dZ-8FuBQFyrkhy:b72fx9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2752)):*
> ```Go
> 		}
> 	}
> 	// Construct an expression which is the sum of all of the casted and hashed
> ```

[nit] comment needs updating ("casted")
___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfBi9iM6fTcVJNih7BX:b-plenb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
> ```Go
> 			expr = &tree.BinaryExpr{
> 				Left:     modBuckets(hashedColumnExpr(i)),
> 				Operator: tree.MakeBinaryOperator(tree.Plus),
> ```

Would it be easier to just do `fnv32(encode_key(col1, col2, col3)) % buckets`?
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfBhAP0Dqaz7S6yrPba:b6raf81) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
> ```Go
> 		// TODO(ajwerner,RaduBerinde): Deal with the cases like
> 		// `crdb_internal.key_encode(f4+1)`.
> 		if funcExpr, ok := e.(*FunctionExpr); ok && funcExpr.Properties.CompositeInsensitive {
> ```

This can be simplified to just make sure that checkTypes is false below when `funcExpr.Properties.CompositeInsensitive = true`. The code below already checks that the inputs are not sensitive.
___
*[pkg/sql/sem/builtins/builtins.go, line 3548 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBdPrD9EQQe1gxaM1M:-MfBdPrD9EQQe1gxaM1N:b-hcgeeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/builtins.go#L3548)):*
> ```Go
> 			},
> 			Volatility: tree.VolatilityImmutable,
> 			Info: "Converts datums into key-encoded bytes. " +
> ```

Not sure if it should be here or in a comment above the built-in but we should mention the property `key_encode(a) == key_encode(b) iff (a IS NOT DISTINCT FROM b)`
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBe-cH1ZufPasyAj4O:-MfBe-cH1ZufPasyAj4P:bfsh1cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L1)):*
> ```Go
> package builtins_test
> ```

[nit] missing header
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBf7_z7MnX1fvdOpiX:-MfBf7_z7MnX1fvdOpiY:b-iu1up7) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L24)):*
> ```Go
> // can cross product them all and we get different values which align to the uniformity
> // expectations.
> func TestCrdbInternalKeyEncode(t *testing.T) {
> ```

I don't quite get what this test is trying to achieve, it seems overly complicated. Is it essentially just cross-checking the distinct count against the distinct key count? Wouldn't we get the same or better by generating a random values clause and doing something like
```sql
with v as (values (1),(2),(3))
  select (select count(distinct v.*) from v) - (select count(distinct key_encode(v.*)) from v);
```
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBe9YX9rbAsxqKywfg:-MfBe9YX9rbAsxqKywfh:b-9qvq9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L73)):*
> ```Go
> 	// Generate a query to cross join all of the values of the two rows
> 	// with also the column ordinal between them.
> 	crossJoinQuery := func() string {
> ```

[nit] would help to show a sample query
___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MfBglP3EfXdHBxf_v0h:b-atgf3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
> ```Go
> 	HasSequenceArguments bool
> 
> 	// CompositeInsensitive indicates that this function does not care about
> ```

CompositeInsensitive indicates that this function returns equal results when evaluated on equal inputs. This is a non-trivial property for composite types which can be equal but not identical (e.g. decimals 1.0 and 1.00). For example, converting a decimal to string is not CompositeInsensitive.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfD8uDo2EKRqJhrsZws:bve3rt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Would it be easier to just do `fnv32(encode_key(col1, col2, col3)) % buckets`?
</blockquote></details>

I was looking at the distribution of values and startled by it which lead me down a small rabbit hole. Consider the test I wrote with the big cross join of two rows. All of the rows end up being unique and constructing unique keys using `encode_key`. What startled me is that it seemed that no matter how I ended up randomizing the inputs, the bucket counts were *perfectly* equal. That alarmed me. It got me thinking that `fnv32` must have strong correlations that are not being taken into account. 

To make this concrete, let's imagine the following for the one non-nil row
```go
	tdb.Exec(t, `INSERT
	INTO t
	SELECT mod(i, 1<<15)::INT2,
		i::INT4,
		i::INT8,
		i::FLOAT4,
		i::FLOAT8,
		i::STRING,
		i::CHAR,
		i::STRING::BYTES,
		i::DECIMAL,
		i::INTERVAL,
		i::OID,
		i::TIMESTAMPTZ,
		i::TIMESTAMP,
		i::DATE,
		('127.0.0.' || mod(i, 256)::STRING)::INET,
		i::VARBIT,
		ARRAY[i::STRING],
		ARRAY[i]
	FROM (
		SELECT $1::INT AS i
	);`, rand.Intn(math.MaxInt32))
```

If I do this query with %s as the cross join (including or not! the column ordinals), I always find the following distribution 
```sql
SELECT k, count(k) AS c
                FROM (
                        SELECTmod(fnv32(crdb_internal.key_encode(input.*)), 8) AS k
                          FROM (%s) AS input
                     )
            GROUP BY k
              ORDER BY k
```

```
        0, 32768
        1, 32768
        2, 32768
        3, 32768
        4, 32768
        5, 32768
        6, 32768
        7, 32768
```

That scared me, I'd expect it to be independent. In fact, most formulations that don't involve the previous step in the hash of the next step don't end up being independent (the current formula is just as broken).  If you use fnv64, it's a bit better but still not perfect. If you use `sha256` and then `fnv32` then you get something that feels truly independent. Also, if you incorporate the result of the previous step in the current step, then it feels independent. I don't know the right tradeoff or the math here. 

So, here's what I'm thinking for the two options:

```sql
SELECT mod(fnv32(sha256(crdb_internal.key_encode(a.i2, 0, b.i4, 1, c.i8, 2, d.f4, 3, e.f8, 4, f.s, 5, g.c, 6, h.b, 7, i.dc, 8, j.ival, 9, k.oid, 10, l.tstz, 11, m.ts, 12, n.da, 13, o.inet, 14, p.vb, 15, q.sa, 16, r.ia, 17))), 8) 
FROM t a, t b, t c, t d, t e, t f, t g, t h, t i, t j, t k, t l, t m, t n, t o, t p, t q, t r;
```

```sql
SELECT
  *
FROM
  (
    SELECT
      mod(
        fnv32(
          crdb_internal.key_encode(
            a.i2,
            0,
            fnv32(
              crdb_internal.key_encode(
                b.i4,
                1,
                fnv32(
                  crdb_internal.key_encode(
                    c.i8,
                    2,
                    fnv32(
                      crdb_internal.key_encode(
                        d.f4,
                        3,
                        fnv32(
                          crdb_internal.key_encode(
                            e.f8,
                            4,
                            fnv32(
                              crdb_internal.key_encode(
                                f.s,
                                5,
                                fnv32(
                                  crdb_internal.key_encode(
                                    g.c,
                                    6,
                                    fnv32(
                                      crdb_internal.key_encode(
                                        h.b,
                                        7,
                                        fnv32(
                                          crdb_internal.key_encode(
                                            i.dc,
                                            8,
                                            fnv32(
                                              crdb_internal.key_encode(
                                                j.ival,
                                                9,
                                                fnv32(
                                                  crdb_internal.key_encode(
                                                    k.oid,
                                                    10,
                                                    fnv32(
                                                      crdb_internal.key_encode(
                                                        l.tstz,
                                                        11,
                                                        fnv32(
                                                          crdb_internal.key_encode(
                                                            m.ts,
                                                            12,
                                                            fnv32(
                                                              crdb_internal.key_encode(
                                                                n.da,
                                                                13,
                                                                fnv32(
                                                                  crdb_internal.key_encode(
                                                                    o.inet,
                                                                    14,
                                                                    fnv32(crdb_internal.key_encode(p.vb, 15, fnv32(crdb_internal.key_encode(q.sa, 16, fnv32(crdb_internal.key_encode(r.ia, 17))))))
                                                                  )
                                                                )
                                                              )
                                                            )
                                                          )
                                                        )
                                                      )
                                                    )
                                                  )
                                                )
                                              )
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        ),
        8
      )
    FROM
      t AS a, t AS b, t AS c, t AS d, t AS e, t AS f, t AS g, t AS h, t AS i, t AS j, t AS k, t AS l, t AS m, t AS n, t AS o, t AS p, t AS q, t AS r
  );
```


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfEvF00EOooEtSvL1Ok:b-7fchkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was looking at the distribution of values and startled by it which lead me down a small rabbit hole. Consider the test I wrote with the big cross join of two rows. All of the rows end up being unique and constructing unique keys using `encode_key`. What startled me is that it seemed that no matter how I ended up randomizing the inputs, the bucket counts were *perfectly* equal. That alarmed me. It got me thinking that `fnv32` must have strong correlations that are not being taken into account. 

To make this concrete, let's imagine the following for the one non-nil row
```go
	tdb.Exec(t, `INSERT
	INTO t
	SELECT mod(i, 1<<15)::INT2,
		i::INT4,
		i::INT8,
		i::FLOAT4,
		i::FLOAT8,
		i::STRING,
		i::CHAR,
		i::STRING::BYTES,
		i::DECIMAL,
		i::INTERVAL,
		i::OID,
		i::TIMESTAMPTZ,
		i::TIMESTAMP,
		i::DATE,
		('127.0.0.' || mod(i, 256)::STRING)::INET,
		i::VARBIT,
		ARRAY[i::STRING],
		ARRAY[i]
	FROM (
		SELECT $1::INT AS i
	);`, rand.Intn(math.MaxInt32))
```

If I do this query with %s as the cross join (including or not! the column ordinals), I always find the following distribution 
```sql
SELECT k, count(k) AS c
                FROM (
                        SELECTmod(fnv32(crdb_internal.key_encode(input.*)), 8) AS k
                          FROM (%s) AS input
                     )
            GROUP BY k
              ORDER BY k
```

```
        0, 32768
        1, 32768
        2, 32768
        3, 32768
        4, 32768
        5, 32768
        6, 32768
        7, 32768
```

That scared me, I'd expect it to be independent. In fact, most formulations that don't involve the previous step in the hash of the next step don't end up being independent (the current formula is just as broken).  If you use fnv64, it's a bit better but still not perfect. If you use `sha256` and then `fnv32` then you get something that feels truly independent. Also, if you incorporate the result of the previous step in the current step, then it feels independent. I don't know the right tradeoff or the math here. 

So, here's what I'm thinking for the two options:

```sql
SELECT mod(fnv32(sha256(crdb_internal.key_encode(a.i2, 0, b.i4, 1, c.i8, 2, d.f4, 3, e.f8, 4, f.s, 5, g.c, 6, h.b, 7, i.dc, 8, j.ival, 9, k.oid, 10, l.tstz, 11, m.ts, 12, n.da, 13, o.inet, 14, p.vb, 15, q.sa, 16, r.ia, 17))), 8) 
FROM t a, t b, t c, t d, t e, t f, t g, t h, t i, t j, t k, t l, t m, t n, t o, t p, t q, t r;
```

```sql
SELECT
  *
FROM
  (
    SELECT
      mod(
        fnv32(
          crdb_internal.key_encode(
            a.i2,
            0,
            fnv32(
              crdb_internal.key_encode(
                b.i4,
                1,
                fnv32(
                  crdb_internal.key_encode(
                    c.i8,
                    2,
                    fnv32(
                      crdb_internal.key_encode(
                        d.f4,
                        3,
                        fnv32(
                          crdb_internal.key_encode(
                            e.f8,
                            4,
                            fnv32(
                              crdb_internal.key_encode(
                                f.s,
                                5,
                                fnv32(
                                  crdb_internal.key_encode(
                                    g.c,
                                    6,
                                    fnv32(
                                      crdb_internal.key_encode(
                                        h.b,
                                        7,
                                        fnv32(
                                          crdb_internal.key_encode(
                                            i.dc,
                                            8,
                                            fnv32(
                                              crdb_internal.key_encode(
                                                j.ival,
                                                9,
                                                fnv32(
                                                  crdb_internal.key_encode(
                                                    k.oid,
                                                    10,
                                                    fnv32(
                                                      crdb_internal.key_encode(
                                                        l.tstz,
                                                        11,
                                                        fnv32(
                                                          crdb_internal.key_encode(
                                                            m.ts,
                                                            12,
                                                            fnv32(
                                                              crdb_internal.key_encode(
                                                                n.da,
                                                                13,
                                                                fnv32(
                                                                  crdb_internal.key_encode(
                                                                    o.inet,
                                                                    14,
                                                                    fnv32(crdb_internal.key_encode(p.vb, 15, fnv32(crdb_internal.key_encode(q.sa, 16, fnv32(crdb_internal.key_encode(r.ia, 17))))))
                                                                  )
                                                                )
                                                              )
                                                            )
                                                          )
                                                        )
                                                      )
                                                    )
                                                  )
                                                )
                                              )
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        ),
        8
      )
    FROM
      t AS a, t AS b, t AS c, t AS d, t AS e, t AS f, t AS g, t AS h, t AS i, t AS j, t AS k, t AS l, t AS m, t AS n, t AS o, t AS p, t AS q, t AS r
  );
```
</blockquote></details>

In that example, we have two random rows in `t` (generated with that Exec statement) and we're mixing and matching all combinations?

When you say "That scared me, I'd expect it to be independent", what do you mean? Just that the counts shouldn't look so perfect?

What feels wrong is that we do mod 8, so we only look at the last few bits of the fnv32, which is not great for a non-cryptographic hash. Using a prime number for the buckets would fix that. I think we could also mod against a bigger prime number before modding down to 8, or shuffle the bits somehow.

I think this can be a problem even for a single column, though I can't find an example.

My concern with your first suggestion is that sha is fairly expensive to calculate. And the second suggestion feels like its obfuscating the underlying problem but not solving it entirely.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfF2XSi1tYdZdlekU6q:b5x4qd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In that example, we have two random rows in `t` (generated with that Exec statement) and we're mixing and matching all combinations?

When you say "That scared me, I'd expect it to be independent", what do you mean? Just that the counts shouldn't look so perfect?

What feels wrong is that we do mod 8, so we only look at the last few bits of the fnv32, which is not great for a non-cryptographic hash. Using a prime number for the buckets would fix that. I think we could also mod against a bigger prime number before modding down to 8, or shuffle the bits somehow.

I think this can be a problem even for a single column, though I can't find an example.

My concern with your first suggestion is that sha is fairly expensive to calculate. And the second suggestion feels like its obfuscating the underlying problem but not solving it entirely.
</blockquote></details>

I guess this example kind-of illustrates your finding:
```
root@127.185.179.142:26257/defaultdb> select *, crdb_internal.key_encode(input.*), fnv32(crdb_internal.key_encode(input.*)), fnv32(crdb_internal.key_encode(input.*)) % 8 from (select t1.a, t2.b, t3.c, t4.d from t t1, t
 t2, t t3, t t4) as input;
  a | b | c | d |                  crdb_internal.key_encode                   |   fnv32    | ?column?
----+---+---+---+-------------------------------------------------------------+------------+-----------
  1 | 1 | 1 | 1 | \211\005?\360\000\000\000\000\000\000\0221\000\001*\002\000 | 2651793876 |        4
  1 | 1 | 1 | 5 | \211\005?\360\000\000\000\000\000\000\0221\000\001*\012\000 | 2517572924 |        4
  1 | 1 | 5 | 1 | \211\005?\360\000\000\000\000\000\000\0225\000\001*\002\000 | 3622176784 |        0
  1 | 1 | 5 | 5 | \211\005?\360\000\000\000\000\000\000\0225\000\001*\012\000 | 3756397736 |        0
  1 | 5 | 1 | 1 | \211\005@\024\000\000\000\000\000\000\0221\000\001*\002\000 |   13520939 |        3
  1 | 5 | 1 | 5 | \211\005@\024\000\000\000\000\000\000\0221\000\001*\012\000 | 4174267283 |        3
  1 | 5 | 5 | 1 | \211\005@\024\000\000\000\000\000\000\0225\000\001*\002\000 |  983903847 |        7
  1 | 5 | 5 | 5 | \211\005@\024\000\000\000\000\000\000\0225\000\001*\012\000 |  849682895 |        7
  5 | 1 | 1 | 1 | \215\005?\360\000\000\000\000\000\000\0221\000\001*\002\000 | 3370022656 |        0
  5 | 1 | 1 | 5 | \215\005?\360\000\000\000\000\000\000\0221\000\001*\012\000 | 3504243608 |        0
  5 | 1 | 5 | 1 | \215\005?\360\000\000\000\000\000\000\0225\000\001*\002\000 | 2378851396 |        4
  5 | 1 | 5 | 5 | \215\005?\360\000\000\000\000\000\000\0225\000\001*\012\000 | 2244630444 |        4
  5 | 5 | 1 | 1 | \215\005@\024\000\000\000\000\000\000\0221\000\001*\002\000 | 3290441751 |        7
  5 | 5 | 1 | 5 | \215\005@\024\000\000\000\000\000\000\0221\000\001*\012\000 | 3156220799 |        7
  5 | 5 | 5 | 1 | \215\005@\024\000\000\000\000\000\000\0225\000\001*\002\000 | 2299270491 |        3
  5 | 5 | 5 | 5 | \215\005@\024\000\000\000\000\000\000\0225\000\001*\012\000 | 2165049539 |        3
```

Note how changing the last value doesn't change the bucket, even though the overall fnv value changes. I think it's probably possible that, given the very specific way in which these rows are constructed, for any row you can find 32768 other rows where the values that are different don't affect the lower 8 bits of the fnv. Or something along those lines.

Not sure how much of an alarm this whole thing is.. It would be more alarming if we could find data where the buckets are super skewed.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfF538W2eR7FeqIZBRy:b27g1r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I guess this example kind-of illustrates your finding:
```
root@127.185.179.142:26257/defaultdb> select *, crdb_internal.key_encode(input.*), fnv32(crdb_internal.key_encode(input.*)), fnv32(crdb_internal.key_encode(input.*)) % 8 from (select t1.a, t2.b, t3.c, t4.d from t t1, t
 t2, t t3, t t4) as input;
  a | b | c | d |                  crdb_internal.key_encode                   |   fnv32    | ?column?
----+---+---+---+-------------------------------------------------------------+------------+-----------
  1 | 1 | 1 | 1 | \211\005?\360\000\000\000\000\000\000\0221\000\001*\002\000 | 2651793876 |        4
  1 | 1 | 1 | 5 | \211\005?\360\000\000\000\000\000\000\0221\000\001*\012\000 | 2517572924 |        4
  1 | 1 | 5 | 1 | \211\005?\360\000\000\000\000\000\000\0225\000\001*\002\000 | 3622176784 |        0
  1 | 1 | 5 | 5 | \211\005?\360\000\000\000\000\000\000\0225\000\001*\012\000 | 3756397736 |        0
  1 | 5 | 1 | 1 | \211\005@\024\000\000\000\000\000\000\0221\000\001*\002\000 |   13520939 |        3
  1 | 5 | 1 | 5 | \211\005@\024\000\000\000\000\000\000\0221\000\001*\012\000 | 4174267283 |        3
  1 | 5 | 5 | 1 | \211\005@\024\000\000\000\000\000\000\0225\000\001*\002\000 |  983903847 |        7
  1 | 5 | 5 | 5 | \211\005@\024\000\000\000\000\000\000\0225\000\001*\012\000 |  849682895 |        7
  5 | 1 | 1 | 1 | \215\005?\360\000\000\000\000\000\000\0221\000\001*\002\000 | 3370022656 |        0
  5 | 1 | 1 | 5 | \215\005?\360\000\000\000\000\000\000\0221\000\001*\012\000 | 3504243608 |        0
  5 | 1 | 5 | 1 | \215\005?\360\000\000\000\000\000\000\0225\000\001*\002\000 | 2378851396 |        4
  5 | 1 | 5 | 5 | \215\005?\360\000\000\000\000\000\000\0225\000\001*\012\000 | 2244630444 |        4
  5 | 5 | 1 | 1 | \215\005@\024\000\000\000\000\000\000\0221\000\001*\002\000 | 3290441751 |        7
  5 | 5 | 1 | 5 | \215\005@\024\000\000\000\000\000\000\0221\000\001*\012\000 | 3156220799 |        7
  5 | 5 | 5 | 1 | \215\005@\024\000\000\000\000\000\000\0225\000\001*\002\000 | 2299270491 |        3
  5 | 5 | 5 | 5 | \215\005@\024\000\000\000\000\000\000\0225\000\001*\012\000 | 2165049539 |        3
```

Note how changing the last value doesn't change the bucket, even though the overall fnv value changes. I think it's probably possible that, given the very specific way in which these rows are constructed, for any row you can find 32768 other rows where the values that are different don't affect the lower 8 bits of the fnv. Or something along those lines.

Not sure how much of an alarm this whole thing is.. It would be more alarming if we could find data where the buckets are super skewed.
</blockquote></details>

*lower 3 bits of the fnv


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfF7_-iBgVhvXfM808s:b-alzjhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

*lower 3 bits of the fnv
</blockquote></details>

Given all of this and efficiency considerations, I'm thinking about just adding a different builtin which takes the datums and hashes them but chains the hash values together to make things . Would such a function upset you? I can also be convinced to stop caring. 
```
	var h fnv.New32()
        var buf []byte
	for i, datum := range args {
                buf = buf[:0]
                var err error
                buf, err = rowenc.EncodeTableKey(out, datum, encoding.Ascending)
                if err != nil { return err }
		_, err := h.Write(buf)
                if err != nil { ... }
                var scratch [4]byte
                binary.PutUvarint(scratch, i) // hash the index hash value to make sequences of the same value independent
                _, err := h.Write(scratch[:])
                if err != nil { ... }
                binary.PutUvarint(scratch, h.Sum32()) // hash the current hash value to increase the data independence
                _, err := h.Write(scratch[:])
                if err != nil { ... }
	}
```
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFBP__-AoL6UgHGzXh:bwwv53b) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This can be simplified to just make sure that checkTypes is false below when `funcExpr.Properties.CompositeInsensitive = true`. The code below already checks that the inputs are not sensitive.
</blockquote></details>

I don't think that's quite true. The `Child(0)` of `FunctionExpr` is its arg list. I think I could do the following but something about it felt icky at the time. 

```go
checkTypes := !opt.IsCompositeInsensitiveOp(e)
if funcExpr, ok := e.(*FunctionExpr); ok && funcExpr.Properties.CompositeInsensitive {
    checkTypes = false
    e = funcExpr.Args
}
```
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBf7_z7MnX1fvdOpiX:-MfFH5qf42SGSBoYb5T0:b-ypmvnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L24)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't quite get what this test is trying to achieve, it seems overly complicated. Is it essentially just cross-checking the distinct count against the distinct key count? Wouldn't we get the same or better by generating a random values clause and doing something like
```sql
with v as (values (1),(2),(3))
  select (select count(distinct v.*) from v) - (select count(distinct key_encode(v.*)) from v);
```
</blockquote></details>

Sure, I was mostly just testing that all the data types worked and that they NULLs worked and then got too clever. I'll take another pass. 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfFIA4S1jrCotl03EMb:bp2nzr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Given all of this and efficiency considerations, I'm thinking about just adding a different builtin which takes the datums and hashes them but chains the hash values together to make things . Would such a function upset you? I can also be convinced to stop caring. 
```
	var h fnv.New32()
        var buf []byte
	for i, datum := range args {
                buf = buf[:0]
                var err error
                buf, err = rowenc.EncodeTableKey(out, datum, encoding.Ascending)
                if err != nil { return err }
		_, err := h.Write(buf)
                if err != nil { ... }
                var scratch [4]byte
                binary.PutUvarint(scratch, i) // hash the index hash value to make sequences of the same value independent
                _, err := h.Write(scratch[:])
                if err != nil { ... }
                binary.PutUvarint(scratch, h.Sum32()) // hash the current hash value to increase the data independence
                _, err := h.Write(scratch[:])
                if err != nil { ... }
	}
```

</blockquote></details>

In general I'm opposed to coming up with this kind of one-off supposed improvements, unless it is following some standard practice or has clear math/reasoning behind it. It's a bit like trying to come up with your own security scheme.

Passing the current hash value through the hasher makes no sense to me. What "data independence" does it increase? It's mathematically equivalent to applying this obscure, static uint32->uint32 function to the current hash. At best, it's a bijective function and you effectively did nothing. At worst, it's collapsing different hashes into the same hash, making things worse.

Also don't see a benefit to hashing the index value (given that the concatenated key is such that you can split it into pieces without ambiguity). I don't know what "make sequences of the same value independent" means.

To me the legit concept is that we take what would be the key for a non-sharded index and hash that. It doesn't get cleaner than that conceptually - it's equivalent to having a hash-based KV store, which is what this is trying to simulate in the first place. I have no objection to a built-in that does that (to save on holding the entire key in memory).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFNVbs0urqedQPKGVD:b-z7holw) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think that's quite true. The `Child(0)` of `FunctionExpr` is its arg list. I think I could do the following but something about it felt icky at the time. 

```go
checkTypes := !opt.IsCompositeInsensitiveOp(e)
if funcExpr, ok := e.(*FunctionExpr); ok && funcExpr.Properties.CompositeInsensitive {
    checkTypes = false
    e = funcExpr.Args
}
```
</blockquote></details>

So? It's a recursive function..


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFTKF90k8DaL_WLgIX:b-ncplm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

So? It's a recursive function..
</blockquote></details>

Ah, maybe ScalarListOp is not marked as insensitive, we should fix that


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2761 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBi9iM6fTcVJNih7BW:-MfFPAB19VZY_8qcyNFX:bw7fnz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2761)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In general I'm opposed to coming up with this kind of one-off supposed improvements, unless it is following some standard practice or has clear math/reasoning behind it. It's a bit like trying to come up with your own security scheme.

Passing the current hash value through the hasher makes no sense to me. What "data independence" does it increase? It's mathematically equivalent to applying this obscure, static uint32->uint32 function to the current hash. At best, it's a bijective function and you effectively did nothing. At worst, it's collapsing different hashes into the same hash, making things worse.

Also don't see a benefit to hashing the index value (given that the concatenated key is such that you can split it into pieces without ambiguity). I don't know what "make sequences of the same value independent" means.

To me the legit concept is that we take what would be the key for a non-sharded index and hash that. It doesn't get cleaner than that conceptually - it's equivalent to having a hash-based KV store, which is what this is trying to simulate in the first place. I have no objection to a built-in that does that (to save on holding the entire key in memory).
</blockquote></details>

I don't have a formalism about the "independence" so I'll stop worrying about it. Anecdotally I observe that if you feed the hash value back in that two rows seem to be no longer related. I've been convinced that I don't care about that. I also don't care much about the buffering. Thanks for all the engagement, I've been convinced to stop caring. 
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFOokg5QNUUJlGoThS:bxsmlme) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, maybe ScalarListOp is not marked as insensitive, we should fix that
</blockquote></details>

The value of `checkTypes` is not passed in recursively (for good reason). The `ScalarListExpr` that is `Args` will have its types checked, right?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfF_4W51_KY-Tl2i9Ot:b99bkqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The value of `checkTypes` is not passed in recursively (for good reason). The `ScalarListExpr` that is `Args` will have its types checked, right?
</blockquote></details>

Not if it (the ScalarListOp) is marked as an insensitive op. We are checking that all paths to variable leaves only pass through composite-insensitive ops.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFaYoG8NjO3B9j5Sfd:b-lmmnki) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not if it (the ScalarListOp) is marked as an insensitive op. We are checking that all paths to variable leaves only pass through composite-insensitive ops.
</blockquote></details>

If `ScalarListOp` were marked as an insensitive op then wouldn't that mean that function invocations be treated as insensitive? At that point anything related to the function properties would be irrelevant. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFc75_3k43I1Sh0x21:b-pu2v3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

If `ScalarListOp` were marked as an insensitive op then wouldn't that mean that function invocations be treated as insensitive? At that point anything related to the function properties would be irrelevant. 
</blockquote></details>

An example of a function that, I think, should be composite sensitive is `json_build_object`. I think that that thing already has some problems:


```
demo@127.0.0.1:52608/movr> SELECT json_build_object('a', -0.0);
  json_build_object
---------------------
  {"a": 0.0}
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

demo@127.0.0.1:52608/movr> SELECT json_build_object('a', -0);
  json_build_object
---------------------
  {"a": 0}
(1 row)


demo@127.0.0.1:52608/movr> SELECT (json_build_object('a', '-0.0'::JSON)->'a')::FLOAT;
  float8
----------
      -0
(1 row)
```


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFdZGz3em--6LElmWg:b-wkhg9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

An example of a function that, I think, should be composite sensitive is `json_build_object`. I think that that thing already has some problems:


```
demo@127.0.0.1:52608/movr> SELECT json_build_object('a', -0.0);
  json_build_object
---------------------
  {"a": 0.0}
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

demo@127.0.0.1:52608/movr> SELECT json_build_object('a', -0);
  json_build_object
---------------------
  {"a": 0}
(1 row)


demo@127.0.0.1:52608/movr> SELECT (json_build_object('a', '-0.0'::JSON)->'a')::FLOAT;
  float8
----------
      -0
(1 row)
```
</blockquote></details>

ScalarListOp is "under" the function. The function would be checked separately.

I believe that what you're discovering above is that you can't simply use -0 in a statement to create a  -0 float (and I believe that's consistent with pg).


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFgnKm1UjDedPWPij0:b-qnv4kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

ScalarListOp is "under" the function. The function would be checked separately.

I believe that what you're discovering above is that you can't simply use -0 in a statement to create a  -0 float (and I believe that's consistent with pg).
</blockquote></details>

Sorry for the confusion. What I was missing was that `getOuterCols` was also recursive. I think that that makes this check N^2 in the depth of the expression but yes, if we also mark `ScalarListOp` as `IsCompositeInsensitiveOp` then it all works out. Thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MfFjZ9f2VpHIHC99Xz-:b-7rehyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Sorry for the confusion. What I was missing was that `getOuterCols` was also recursive. I think that that makes this check N^2 in the depth of the expression but yes, if we also mark `ScalarListOp` as `IsCompositeInsensitiveOp` then it all works out. Thanks!
</blockquote></details>

Good point, maybe leave a TODO in there for the quadratic behavior.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I noticed there already is a `crdb_internal.encode_key`. I don't think we can use that because it's meant to encode a full key, but we should rename the new one so there's no confusion between the two.

RaduBerinde reviewed (commented):
Looking good!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-:-MhU1r-s7AvYPJgWkK7y:br2sbr6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2569 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MhU0QkD8YSdlPDxoRBk:-MhU0QkD8YSdlPDxoRBl:bfg5gtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4279e15ef8c11830caf66fe2b2616fb736dbb799/pkg/sql/create_table.go#L2569)):*
> ```Go
> // makeShardColumnDesc returns a new column descriptor for a hidden computed shard column
> // based on all the `colNames`.
> func makeShardColumnDesc(
> ```

[nit] this deserves a comment explaining the argument, maybe pointing to the two make*ComputeExpr for details.
___
*[pkg/sql/create_table.go, line 2625 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MhU0xUCCiA_B6ZvY24I:-MhU0xUCCiA_B6ZvY24J:b-20bh7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/4279e15ef8c11830caf66fe2b2616fb736dbb799/pkg/sql/create_table.go#L2625)):*
> ```Go
> 			Exprs: tree.Exprs{
> 				expr,
> 				&tree.CastExpr{
> ```

Is this cast necessary? Maybe we can just use `tree.AdjustValueToType` instead? Though I think we should error out much earlier if the value is too large.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/create_table.go, line 2716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhnzQ-dhs_EYUS0au:-MhUoVXJ0Qc_oRoNh1q3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/create_table.go#L2716)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] probably on your TODO list but this should have a comment with sample expressions
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 2569 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MhU0QkD8YSdlPDxoRBk:-MhV-kEmB3TDArLDAhgl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4279e15ef8c11830caf66fe2b2616fb736dbb799/pkg/sql/create_table.go#L2569)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this deserves a comment explaining the argument, maybe pointing to the two make*ComputeExpr for details.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 2625 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MhU0xUCCiA_B6ZvY24I:-MhUt7js1vzXjuZZv_Ck:b-3d8mq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4279e15ef8c11830caf66fe2b2616fb736dbb799/pkg/sql/create_table.go#L2625)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Is this cast necessary? Maybe we can just use `tree.AdjustValueToType` instead? Though I think we should error out much earlier if the value is too large.
</blockquote></details>

Removed. Also added a check and test. 
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 2644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBhAP0Dqaz7S6yrPb_:-MhU-X1P0mX9D1IIKhAj:b-xl7g2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/opt/memo/logical_props_builder.go#L2644)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good point, maybe leave a TODO in there for the quadratic behavior.
</blockquote></details>

This is now fixed :tada: 
___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MhU-ZJl06skJRkjB_-_:b-gvztkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

CompositeInsensitive indicates that this function returns equal results when evaluated on equal inputs. This is a non-trivial property for composite types which can be equal but not identical (e.g. decimals 1.0 and 1.00). For example, converting a decimal to string is not CompositeInsensitive.
</blockquote></details>

Removed as it's not needed anymore. 
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBe-cH1ZufPasyAj4O:-MhU-T4a86RsLyG5a6m6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] missing header
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBf7_z7MnX1fvdOpiX:-MhU-TzW4kBsxKKZJBz9:b184x9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L24)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Sure, I was mostly just testing that all the data types worked and that they NULLs worked and then got too clever. I'll take another pass. 
</blockquote></details>

Cleaned it up roughly as you suggested.
___
*[pkg/sql/sem/builtins/key_encode_builtin_test.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBe9YX9rbAsxqKywfg:-MhU-W0nEoIDoGEFJNDP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b422326d43c579c450ccde6606b30dc67e75fa/pkg/sql/sem/builtins/key_encode_builtin_test.go#L73)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] would help to show a sample query
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MhV89liD7xp29MurrDq:b-fqhceo) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Removed as it's not needed anymore. 
</blockquote></details>

Hm, how come? I don't think we can infer a hash value for composite types without it..


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MhVAa4x5grPS53eCWNg:bcx3y0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, how come? I don't think we can infer a hash value for composite types without it..
</blockquote></details>

I think you might have encountered a bug. You made the scalar list composite insensitive and we're probably hitting this code:
```
		if opt.IsCompositeInsensitiveOp(e) {
			// The operator is known to be composite-insensitive. If its result is a
			// non-composite type, it is also composite-identical.
			return true, !colinfo.HasCompositeKeyEncoding(e.(opt.ScalarExpr).DataType())
		}
```
The data type of scalar list is `Any` and it's not handled. There are others like `Tuple` that aren't handled. We should change `HasCompositeKeyEncoding` to `CanHaveCompositeKeyEncoding` and have the default return value for unhandled types be true.


<!-- Sent from Reviewable.io -->

ajwerner committed:
colinfo: rename HasCompositeKeyEncoding to CanHaveCompositeKeyEncoding

This patch also fixes the behavior to properly catagorize types like any
and unknown. It does so by adding all of the families and setting the default
to be true instead of the prior false.

Release note: None

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MhVTKO29HUXqg_Xl-I6:b-n4klxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think you might have encountered a bug. You made the scalar list composite insensitive and we're probably hitting this code:
```
		if opt.IsCompositeInsensitiveOp(e) {
			// The operator is known to be composite-insensitive. If its result is a
			// non-composite type, it is also composite-identical.
			return true, !colinfo.HasCompositeKeyEncoding(e.(opt.ScalarExpr).DataType())
		}
```
The data type of scalar list is `Any` and it's not handled. There are others like `Tuple` that aren't handled. We should change `HasCompositeKeyEncoding` to `CanHaveCompositeKeyEncoding` and have the default return value for unhandled types be true.
</blockquote></details>

Alrighty, it's back and it's now confirmed needed. See how you feel about the new logic for `CanHaveCompositeKeyEncoding`. 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/sem/tree/function_definition.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-MfBglP3EfXdHBxf_v0g:-MhVWM-66vmkQNZEsaOz:bb1m50l) ([raw file](https://github.com/cockroachdb/cockroach/blob/667a4129bc8cc90cda01a6b2fd93f59d86e793a8/pkg/sql/sem/tree/function_definition.go#L108)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Alrighty, it's back and it's now confirmed needed. See how you feel about the new logic for `CanHaveCompositeKeyEncoding`. 
</blockquote></details>

:thumbsup: :thumbsup: looks great!


<!-- Sent from Reviewable.io -->

ajwerner commented:
@RaduBerinde anything else you want on this?

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Just wrap it with a red bow!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67865#-:-MhZJz8TCwYlcFSIdpx7:bltkxmn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
